core/templates/pages/exploration-player-page/services/command-executor.service.ts

```diff
@@ -0,0 +1,210 @@
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Oppia interaction from outer iframe with messaging commands.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import {WindowRef} from 'services/contextual/window-ref.service.ts';
+import {WindowWrapperMessageService} from
+  'pages/exploration-player-page/services/window-wrapper-message.service';
+require('pages/exploration-player-page/services/exploration-engine.service.ts');
+import {ServicesConstants} from 'services/services.constants';
+
+@Injectable({providedIn: 'root'})
+export class CommandExecutorService {
+  private commandToFunctionMap = {
+    CONTINUE: this.continueClick,
+    ENTER_TEXT_NUMBER_UNITS: this.fillTextBox,
+    ADD_SET: this.addSet,
+    REMOVE_SET: this.removeSet,
+    ENTER_FRACTION: this.enterFraction,
+    SELECT_ITEM_BULLET: this.selectItemBullet,
+    SUBMIT: this.submit,
+  };
+  setElementsOnPage = 0;
+  hostname = '';
+  cachedOuterFrameMessage = '';
+  windowWrapperMessageService = null;
+  constructor(private windowRef: WindowRef) {
+    this.setElementsOnPage = 0;
+    this.hostname = '';
+    this.cachedOuterFrameMessage = '';
+    this.windowWrapperMessageService =
+      new WindowWrapperMessageService(windowRef);
+  }
+
+  sendParentReadyState(windowRef) {
+    this.windowWrapperMessageService.postMessageToParent(
+      'Ready to receive hostname', '*');
+  }
+
+  getOuterFrameEvents(windowRef) {
+    this.windowWrapperMessageService.addEventListener(
+      'message', (event) => {
+        var messageArray = event.data.split(' ');
+        var command = messageArray[0];
+        var message = '';
+        for (var i = 1; i < messageArray.length; i++) {
+          message = message + messageArray[i] + ' ';
+        }
+        message = message.substr(0, message.length - 1);
+        if ((command !== 'HOSTNAME' && this.hostname === '') ||
+        command === 'finishedProcessing') {
+          return;
+        } else if (command === 'CONTINUE' || command === 'SUBMIT') {
+          this.commandToFunctionMap[command](windowRef);
+        } else if (command === 'HOSTNAME') {
+          if (ServicesConstants.WHITELISTED_IFRAME_HOSTS.indexOf(message) >=
+          0) {
+            this.hostname = message;
+          }
+          if (this.cachedOuterFrameMessage !== '' &&
+          this.cachedOuterFrameMessage !== undefined) {
+            this.windowWrapperMessageService.postMessageToParent(
+              this.cachedOuterFrameMessage, this.hostname);
+          }
+        } else {
+          this.commandToFunctionMap[command](windowRef, message);
+        }
+      });
+  }
+
+  sendStateToOuterFrame(id) {
```

Comment:
You'll need to think about versioning here. Over time, the protocol may change, or the format of things we send may be different, etc. 

What's the plan for handling that? There should at least be a version number for the protocol. But it's also important to think through how the upgrade procedure, as well as the deprecation of older protocols, is going to work -- we can't promise unlimited backwards-compatibility.